### PR TITLE
enhance logs functionality to handle build stage log

### PR DESCRIPTION
**Pull request by James-JLC**
Signed-off-by: Alex Ji <jilichao@aliyun.com>

### CODE REVIEW

File: pkg/cmd/subcommand/logs.go [hunk 1]
```diff
@@ -111,31 +114,114 @@ func (l *Logs) preRun(cf *genericclioptions.ConfigFlags, args []string) error {
...
+		var logReader io.ReadCloser
+		if l.containerName != "" {
+			logReader, err = podInterface.GetLogs(pod.Name, &corev1.PodLogOptions{Follow: l.Follow, Container: l.containerName}).Stream(ctx)
+			if err != nil {
+				return err
+			}
+			defer logReader.Close()
+			readerList = append(readerList, logReader)
+		} else {
+			for _, container := range pod.Spec.Containers {
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/logs.go:192]: There're several kinds of containers in a Function pod:
- the function container whose name is just `function`
- the dapr sidecar container
- the knative sidecar container

Normally we only output logs from the `function` container is enough. This is why we set container name to `function` in https://github.com/OpenFunction/cli/pull/48/commits/21bea4c4b7017c1d728ddb6943cfc942ac8cd154#diff-5c7fb0ae4b5f3afe942a2e3eed411e5a10ffff3e1dccbbf969947d98b79d60d7L54

If the user needs to view logs from dapr sidecar or knative sidecar, he can specify container name. If container name is not specified, the default one `function` will be used.

So we needn't to add a loop here to iterate all containers

File: pkg/cmd/subcommand/logs.go [hunk 2]
```diff
@@ -111,31 +114,114 @@ func (l *Logs) preRun(cf *genericclioptions.ConfigFlags, args []string) error {
...
+		}
+
+		// get shipwright builder-buildrun
+		swbuildrunRef := builder.Status.ResourceRef["shipwright.io/buildRun"]
+		swBuildRun, err := l.swClient.ShipwrightV1alpha1().BuildRuns(f.Namespace).Get(ctx, swbuildrunRef, metav1.GetOptions{})
+		if err != nil {
+			statusError, ok := err.(*k8serrors.StatusError)
+			// buildrun has been cleaned up
+			if ok && statusError.Status().Code == 404 {
+				goto serving
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/logs.go]: The `goto serving` logic here is to fetch serving logs continually after the build phase ends.

But I think we can make it simpler by removing the 'goto serving' logic:
- If there is builder CR exists, just continually fetch build logs by adding `-f` option. The builder CR will not be deleted if the build fails and will be deleted when the build succeeds.
- If there is no builder CR, and there is serving CR. This means the build is successful, so we only need to fetch serving logs
- If there is no builder CR and no serving CR either, means this is in between the build and serving phase, and we needn't fetch any logs. And we needn't wait for the serving pod to be started either. The possibility of this scenario is small.

In short, we only need to fetch logs from the currently existing pod's function container by default.
No need to wait for the serving pod to up and running.

Comment (James-JLC) [reply pkg/cmd/subcommand/logs.go]: Yes, I totally agree with you! I thought too much before.

Thanks for your review!

**Comment by benjaminhuo** [comment]
@jilichao Thanks for the PR. This is a big enhancement and we're going to release a new version after this PR!

**Review by benjaminhuo — APPROVED** [review]
